Swap out `ethers-eip712`

In #213 I got rid of the full, bloaty `ethers` module in favor of smaller packages that we actually use. Unfortunately, I have found that `ethers-eip712`, which we need for TypeData messages, requires the full `ethers` module as a dependency, so this SDK will break if that is not included.

I have looked at several EIP712 libs but none of them work for the test payloads so I can't use any of them right now. It isn't worth the trouble of fixing those other modules, so I will include `ethers` back in for now and come back to this.

The best option I found is a module called [`eip-712`](https://www.npmjs.com/package/eip-712). It failed on several payloads, but it looked like only two types of failures:

* `'Should test a large 1inch transaction'` threw some error internal to `eip-712`: `AssertionError: expected 'Failed to make request to device: data.startsWith is not a function' to equal null`
* `'Should test canonical EIP712 example with 3rd level nesting'` simply calculates the wrong hash. I opened [an issue](https://github.com/Mrtenz/eip-712/issues/34#issuecomment-1006863366) pointing this out. I looked at the code and it is not obvious what the issue is except to say that it is failing to build the final nested object into the definition. The `encodeType` function in `ethers-eip712` ([here](https://github.com/0xsequence/ethers-eip712/blob/master/src/typed-data.ts#L122)) is much more complex than the one in `eip-712` ([here](https://github.com/Mrtenz/eip-712/blob/master/src/eip-712.ts#L50)) so I think that code would probably just need to get replicated. Unfortunately that isn't the only problem so I can't just fork it and merge that in.

We should aim to get rid of `ethers-eip712` once we find a slimmer module that works with all the payloads we need.

I have started a branch to show what this swap could look like: https://github.com/GridPlus/gridplus-sdk/tree/fix-ethers-again

--- a/package.json
+++ b/package.json
@@ -24,7 +24,6 @@
     "url": "https://github.com/GridPlus/gridplus-sdk.git"
   },
   "dependencies": {
-    "@ethersproject/bignumber": "^5.5.0",
     "aes-js": "^3.1.1",
     "bech32": "^2.0.0",
     "bignumber.js": "^9.0.1",
@@ -34,8 +33,7 @@
     "buffer": "^5.6.0",
     "crc-32": "^1.2.0",
     "elliptic": "6.5.4",
-    "ethers": "^5.5.2",
-    "ethers-eip712": "^0.2.0",
+    "eth-eip712-util": "^2.3.1",
     "js-sha3": "^0.8.0",
     "rlp-browser": "^1.0.1",
     "secp256k1": "4.0.2",

--- a/src/ethereum.js
+++ b/src/ethereum.js
@@ -4,8 +4,7 @@ const BN = require('bignumber.js');
 const Buffer = require('buffer/').Buffer;
 const cbor = require('borc');
 const constants = require('./constants');
-const ethersBigNum = require('@ethersproject/bignumber');
-const eip712 = require('ethers-eip712');
+const eip712 = require('eth-eip712-util').TypedDataUtils;
 const keccak256 = require('js-sha3').keccak256;
 const rlp = require('rlp-browser');
 const secp256k1 = require('secp256k1');
@@ -49,7 +48,8 @@ exports.validateEthereumMsgResponse = function(res, req) {
     // Get recovery param with a `v` value of [27,28] by setting `useEIP155=false`
     return addRecoveryParam(hash, sig, signer, { chainId: 1, useEIP155: false })
   } else if (input.protocol === 'eip712') {
-    const digest = prehash ? prehash : eip712.TypedDataUtils.encodeDigest(req.input.payload);
+    const encoded = eip712.hash(req.input.payload);
+    const digest = prehash ? prehash : encoded
     // Get recovery param with a `v` value of [27,28] by setting `useEIP155=false`
     return addRecoveryParam(digest, sig, signer, { useEIP155: false })
   } else {
@@ -650,9 +650,9 @@ function buildEIP712Request(req, input) {
       throw new Error('EIP712Domain type must be defined.')
     // Parse the payload to ensure we have valid EIP712 data types and that
     // they are encoded such that Lattice firmware can parse them.
-    // We need two different encodings:
-    // 1. Use `ethers` BigNumber when building the request to be validated by ethers-eip712.
-    //    Make sure we use a copy of the data to avoid mutation problems
+    // We need two different encodings: one to send to the Lattice in a format that plays
+    // nicely with our firmware CBOR decoder. The other is formatted to be consumable by
+    // our EIP712 validation module.
     input.payload.message = parseEIP712Msg( JSON.parse(JSON.stringify(data.message)), 
                                             JSON.parse(JSON.stringify(data.primaryType)), 
                                             JSON.parse(JSON.stringify(data.types)), 
@@ -661,8 +661,6 @@ function buildEIP712Request(req, input) {
                                             'EIP712Domain', 
                                             JSON.parse(JSON.stringify(data.types)), 
                                             true);
-    // 2. Use `bignumber.js` for the request going to the Lattice, since it's the required
-    //    BigNumber lib for `cbor`, which we use to encode the request data to the Lattice.
     data.domain = parseEIP712Msg(data.domain, 'EIP712Domain', data.types, false);
     data.message = parseEIP712Msg(data.message, data.primaryType, data.types, false);
     // Now build the message to be sent to the Lattice
@@ -672,7 +670,8 @@ function buildEIP712Request(req, input) {
     if (fwConst.ethMsgPreHashAllowed && payload.length > maxSzAllowed) {
       // If this payload is too large to send, but the Lattice allows a prehashed message, do that
       req.payload.writeUInt16LE(payload.length, off); off += 2;
-      const prehash = Buffer.from(keccak256(eip712.TypedDataUtils.encodeDigest(req.input.payload), 'hex'), 'hex');
+      const encoded = eip712.hash(req.input.payload)
+      const prehash = Buffer.from(keccak256(encoded, 'hex'), 'hex');
       prehash.copy(req.payload, off);
       req.prehash = prehash;
     } else {
@@ -741,7 +740,7 @@ function splitFrames(data, frameSz) {
   return frames;
 }
 
-function parseEIP712Msg(msg, typeName, types, isEthers=false) {
+function parseEIP712Msg(msg, typeName, types, forJSParser=false) {
   try {
     const type = types[typeName];
     type.forEach((item) => {
@@ -755,11 +754,11 @@ function parseEIP712Msg(msg, typeName, types, isEthers=false) {
         // elementary (i.e. non-custom) type.
         // For arrays, we need to loop through each message item.
         for (let i = 0; i < msg.length; i++) {
-            msg[i][item.name] = parseEIP712Msg(msg[i][item.name], singularType, types, isEthers)
+            msg[i][item.name] = parseEIP712Msg(msg[i][item.name], singularType, types, forJSParser)
         }
       } else if (isCustomType) {
         // Not an array means we can jump directly into the sub-struct to convert
-        msg[item.name] = parseEIP712Msg(msg[item.name], singularType, types, isEthers)
+        msg[item.name] = parseEIP712Msg(msg[item.name], singularType, types, forJSParser)
       } else if (Array.isArray(msg)) {
         // If we have an array for this particular type and the type we are parsing
         // is *not* a custom type, loop through the array elements and convert the types.
@@ -769,22 +768,22 @@ function parseEIP712Msg(msg, typeName, types, isEthers=false) {
             // This code is not reachable for custom types so we assume these are arrays of
             // elementary types.
             for (let j = 0; j < msg[i][item.name].length; j++) {
-              msg[i][item.name][j] = parseEIP712Item(msg[i][item.name][j], singularType, isEthers)
+              msg[i][item.name][j] = parseEIP712Item(msg[i][item.name][j], singularType, forJSParser)
             }
           } else {
             // Non-arrays parse + replace one value for the elementary type
-            msg[i][item.name] = parseEIP712Item(msg[i][item.name], singularType, isEthers)
+            msg[i][item.name] = parseEIP712Item(msg[i][item.name], singularType, forJSParser)
           }
         }
       } else if (isArrayType) {
         // If we have an elementary array type and a non-array message level, 
         //loop through the array and parse + replace  each item individually.
         for (let i = 0; i < msg[item.name].length; i++) {
-          msg[item.name][i] = parseEIP712Item(msg[item.name][i], singularType, isEthers)
+          msg[item.name][i] = parseEIP712Item(msg[item.name][i], singularType, forJSParser)
         }
       } else {
         // If this is a singular elementary type, simply parse + replace.
-        msg[item.name] = parseEIP712Item(msg[item.name], singularType, isEthers)
+        msg[item.name] = parseEIP712Item(msg[item.name], singularType, forJSParser)
       }
       
     })
@@ -794,16 +793,24 @@ function parseEIP712Msg(msg, typeName, types, isEthers=false) {
   return msg;
 }
 
-function parseEIP712Item(data, type, isEthers=false) {
+function parseEIP712Item(data, type, forJSParser=false) {
   if (type === 'bytes') {
     // Variable sized bytes need to be buffer type
     data = ensureHexBuffer(data);
+    if (forJSParser) {
+        // For EIP712 encoding module it's easier to encode hex strings
+        data = `0x${data.toString('hex')}`
+    }
   } else if (type.slice(0, 5) === 'bytes') {
     // Fixed sizes bytes need to be buffer type. We also add some sanity checks.
     const nBytes = parseInt(type.slice(5));
     data = ensureHexBuffer(data);
     if (data.length !== nBytes)
       throw new Error(`Expected ${type} type, but got ${data.length} bytes`);
+    if (forJSParser) {
+        // For EIP712 encoding module it's easier to encode hex strings
+        data = `0x${data.toString('hex')}`
+    }
   } else if (type === 'address') {
     // Address must be a 20 byte buffer
     data = ensureHexBuffer(data);
@@ -813,8 +820,8 @@ function parseEIP712Item(data, type, isEthers=false) {
     }
     if (data.length !== 20)
       throw new Error(`Address type must be 20 bytes, but got ${data.length} bytes`);
-    // Ethers wants addresses as hex strings
-    if (isEthers === true) {
+    // For EIP712 encoding module it's easier to encode hex strings
+    if (forJSParser) {
       data = `0x${data.toString('hex')}`
     }
   } else if ( (constants.ethMsgProtocol.TYPED_DATA.typeCodes[type]) && 
@@ -825,9 +832,9 @@ function parseEIP712Item(data, type, isEthers=false) {
       b = Buffer.from('00', 'hex');
     }
     // Uint256s should be encoded as bignums.
-    if (isEthers === true) {
-      // `ethers` uses their own BigNumber lib
-      data = ethersBigNum.BigNumber.from(`0x${b.toString('hex')}`)
+    if (forJSParser) {
+      // For EIP712 encoding in this module we need strings to represent the numbers
+      data = `0x${b.toString('hex')}`
     } else {
       // `bignumber.js` is needed for `cbor` encoding, which gets sent to the Lattice and plays
       // nicely with its firmware cbor lib.

--- a/test/testEthMsg.js
+++ b/test/testEthMsg.js
@@ -218,6 +218,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -257,6 +258,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -308,6 +310,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -325,6 +328,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -368,6 +372,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -421,6 +426,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -484,6 +490,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -557,6 +564,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -630,6 +638,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -691,6 +700,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -764,6 +774,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -822,6 +833,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -910,6 +922,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -995,11 +1008,152 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
+      expect(err).to.equal(null)
+    }
+  })
+
+  it('Should test a payload with a nested type in multiple nesting levels', async () => {   
+    const msg = {
+      'types': {
+        'EIP712Domain': [
+          {
+            'name': 'name',
+            'type': 'string'
+          },
+        ],
+        'PrimaryType': [
+          {
+            'name': 'one',
+            'type': 'Type1'
+          },
+          {
+            'name': 'zero',
+            'type': 'Type0'
+          },
+        ],
+        'Type1': [
+          {
+            'name': '1s',
+            'type': 'string'
+          },
+        ],
+        'Type0': [
+          {
+            'name': 'one',
+            'type': 'Type1'
+          }
+        ]
+      },
+      'primaryType': 'PrimaryType',
+      'domain': {
+        'name': 'Domain',
+      },
+      'message': {
+        'one': {
+          '1s': 'nestedOne',
+        },
+        'zero': {
+          'one': {
+            '1s': 'nestedTwo',
+          }
+        },
+      }
+    }
+
+    const req = {
+      currency: 'ETH_MSG',
+      data: {
+        signerPath: [helpers.BTC_LEGACY_PURPOSE, helpers.ETH_COIN, HARDENED_OFFSET, 0, 0],
+        protocol: 'eip712',
+        payload: msg,
+      }
+    }
+    try {
+      await helpers.execute(client, 'sign', req);
+    } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
 
   it('Should test a payload that requires use of extraData frames', async () => {
+    const msg = {
+      types: {
+        EIP712Domain: [
+          { name: 'name', type: 'string' },
+          { name: 'version', type: 'string' },
+          { name: 'chainId', type: 'uint256' },
+          { name: 'verifyingContract', type: 'address' }
+        ],
+        Wallet: [
+          { name: 'address', type: 'address' },
+          { name: 'balance', type: 'Balance' },
+        ],
+        Person: [
+          { name: 'name', type: 'string' },
+          { name: 'wallet', type: 'Wallet' }
+        ],
+        Mail: [
+          { name: 'from', type: 'Person' },
+          { name: 'to', type: 'Person' },
+          { name: 'contents', type: 'string' }
+        ],
+        Balance: [
+          { name: 'value', type: 'uint256' },
+          { name: 'currency', type: 'string' }
+        ]
+      },
+      primaryType: 'Mail',
+      domain: {
+        name: 'Ether Mail',
+        version: '1',
+        chainId: 12,
+        verifyingContract: '0xCcCCccccCCCCcCCCCCCcCcCccCcCCCcCcccccccC'
+      },
+      message: {
+        from: {
+          name: 'Cow',
+          wallet: {
+            address: '0xCD2a3d9F938E13CD947Ec05AbC7FE734Df8DD826',
+            balance: {
+              value: '0x12345678',
+              currency: 'ETH',
+            }
+          },
+        },
+        to: {
+          name: 'Bob',
+          wallet: {
+            address: '0xbBbBBBBbbBBBbbbBbbBbbbbBBbBbbbbBbBbbBBbB',
+            balance: {
+              value: '0xabcdef12',
+              currency: 'UNI',
+            },
+          },
+        },
+        contents: 'stupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimesstupidlylongstringthatshouldstretchintomultiplepageswhencopiedmanytimes'
+      }
+    };
+    const req = {
+      currency: 'ETH_MSG',
+      data: {
+        signerPath: [helpers.BTC_LEGACY_PURPOSE, helpers.ETH_COIN, HARDENED_OFFSET, 0, 0],
+        protocol: 'eip712',
+        payload: msg,
+      }
+    }
+    try {
+      await helpers.execute(client, 'sign', req);
+    } catch (err) {
+      foundError = true;
+      expect(err).to.equal(null)
+    }
+  });
+
+  it('Should test random edge case #1', async () => {
+    // This was a randomly generated payload which caused an edge case.
+    // It has been slimmed down but definition structure is preserved.
     const msg = {
       'types': {
         'EIP712Domain': [
@@ -1020,95 +1174,61 @@ describe('Test ETH EIP712', function() {
             'type': 'address'
           }
         ],
-        'Primary_Promote_room_donate_': [
+        'Primary_Click': [
           {
-            'name': 'assist_dumb_bubble_g',
-            'type': 'Open_either_fetch_gr'
+            'name': 'utility',
+            'type': 'Expose'
           },
           {
-            'name': 'jungle_symptom_meat_',
-            'type': 'Unfair_device_vocal_'
+            'name': 'aisle',
+            'type': 'Cancel'
           },
           {
-            'name': 'buzz_attract_crater_',
-            'type': 'Capital_junk_pet_dig'
+            'name': 'gym',
+            'type': 'Razor'
           },
           {
-            'name': 'achieve_twenty_found',
-            'type': 'bytes24'
-          },
-          {
-            'name': 'tent_cradle_bamboo_s',
-            'type': 'bytes19'
+            'name': 'drift_patch_cable_bi',
+            'type': 'bytes1'
           }
         ],
-        'Open_either_fetch_gr': [
+        'Expose': [
           {
-            'name': 'afraid_three_silver_',
-            'type': 'string'
-          },
+            'name': 'favorite',
+            'type': 'bytes21'
+          }
+        ],
+        'Cancel': [
           {
-            'name': 'two_wedding_author_a',
+            'name': 'clever',
+            'type': 'uint200'
+          }
+        ],
+        'Razor': [
+          {
+            'name': 'private',
             'type': 'bytes2'
-          },
-          {
-            'name': 'sudden_tail_exclude_',
-            'type': 'bytes11'
-          }
-        ],
-        'Unfair_device_vocal_': [
-          {
-            'name': 'feel_affair_curve_av',
-            'type': 'bytes3'
-          },
-          {
-            'name': 'deer_spoil_indicate_',
-            'type': 'address'
-          }
-        ],
-        'Capital_junk_pet_dig': [
-          {
-            'name': 'return_palace_flock_',
-            'type': 'bytes32'
-          },
-          {
-            'name': 'manage_cement_area_t',
-            'type': 'bytes3'
-          },
-          {
-            'name': 'pole_pluck_odor_circ',
-            'type': 'Open_either_fetch_gr'
           }
         ]
       },
-      'primaryType': 'Primary_Promote_room_donate_',
+      'primaryType': 'Primary_Click',
       'domain': {
-        'name': 'Domain_Census_liberty_own_s',
+        'name': 'Domain_Avocado_luggage_twel',
         'version': '1',
-        'chainId': '0x2207',
-        'verifyingContract': '0x52a3e01d76d2670f8fd452564b3f56eea6fc798d'
+        'chainId': '0x324e',
+        'verifyingContract': '0x69f758a7911448c2f7aa6df15ca27d69ffa1c6b7'
       },
       'message': {
-        'assist_dumb_bubble_g': {
-          'afraid_three_silver_': 'duck_acquire_chaos_rough_leader_merry_symptom_slab_tooth_bachelor_news_produce_bleak_young_skin_toot',
-          'two_wedding_author_a': '0x9604',
-          'sudden_tail_exclude_': '0xf996bed91579769e5bd995',
+        'utility': {
+          'favorite': '0x891b56dc6ab87ab73cf69761183d499283f1925871'
         },
-        'jungle_symptom_meat_': {
-          'feel_affair_curve_av': '0x4f493d',
-          'deer_spoil_indicate_': '0x344b29a452b79bb8e4ef37f2c7688faafd0d2c7d'
+        'aisle': {
+          'clever': '0x0102'
         },
-        'buzz_attract_crater_': {
-          'return_palace_flock_': '0x93012c3a0c21d9adaaa3fb009f06a7bd90449b13df99096f1bdce9c48e16dbec',
-          'manage_cement_area_t': '0x37b91f',
-          'pole_pluck_odor_circ': {
-            'afraid_three_silver_': 'author_never_range_boring_rabbit_meat_notable_excuse_attract_project_east_film_stay_twenty_cause_squ',
-            'two_wedding_author_a': '0xf301',
-            'sudden_tail_exclude_': '0xb5b5f8c2b5120c8aba2497',
-          }
+        'gym': {
+          'private': '0xbb42',
         },
-        'achieve_twenty_found': '0xa5e8c0daacbf617191ec37b0cea3e23b15e4237935733d8c',
-        'tent_cradle_bamboo_s': '0x4bd62f055b4a0ac7f52fe604bd7da4debdd291',
+        'drift_patch_cable_bi': '0xb4'
       }
     }
     const req = {
@@ -1122,6 +1242,7 @@ describe('Test ETH EIP712', function() {
     try {
       await helpers.execute(client, 'sign', req);
     } catch (err) {
+      foundError = true;
       expect(err).to.equal(null)
     }
   })
@@ -1136,5 +1257,5 @@ describe('Test ETH EIP712', function() {
       setTimeout(() => { next(err) }, 2000);
     }
   })
-
+  
 })

